### PR TITLE
(towards #2543) Refresh PSyKAl documentation

### PR DESCRIPTION
**Pull request by sergisiso**
In this PR I have focused on the PSyKAl part of the documentation. I often found it was not very approachable because terms where spread in multiple locations (due to being the initial documentation and new parts added over time). So I have tried to condense the Intro(which was really PSyKAl)/Algorithm/Kernel/PSy/DistributedMem into a single page `introduction_to_psykal.rst`, by removing lots of duplication and making it more concise. Let me know what you think and if I should restore something.

Other than than, I also move a couple of sections to the developers guide, unindexed the NEMO PSyKAl for now (I plan a new page for it), and move all PSyKAl section after the Code Transformation mode section.

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter This is ready for first review.

File: doc/user_guide/introduction_to_psykal.rst [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+.. _built-ins:
+
+Built-ins
+---------
+
+Built-ins are operations which can be specified within an invoke call in
+the algorithm layer but do not require an associated kernel implementation
+because they are provided by the infrastructure.
+
+These are useful for commonly used operations, as they reduce the amount of
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:277]: "commonly-used"

File: doc/user_guide/introduction_to_psykal.rst [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+          same scope. For example, it is not valid to make use of a
+          Built-in called ``setval_c`` and for its parent subroutine
+          to also be called ``setval_c``. In this case PSyclone will
+          raise an exception.
+
+Example
++++++++
+
+In the following algorithm-layer example from the LFRic API, the invoke call includes a call
+to two Built-ins (``setval_c`` and ``X_divideby_Y``) and a user-supplied kernel
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:300]: As Oakley is currenly implementing support for user-supplied kernels that iterate over dofs, we should change this a bit. Perhaps "user-supplied kernel that operates on cell columns..."?

File: doc/user_guide/introduction_to_psykal.rst [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+    ...
+
+    subroutine jacobi_solver_algorithm(lhs, rhs, mm, mesh, n_iter)
+
+      integer(kind=i_def), intent(in)    :: n_iter
+      type(field_type),    intent(inout) :: lhs
+      type(field_type),    intent(in)    :: rhs
+      type(operator_type), intent(in)    :: mm
+      type(mesh_type),     intent(in)    :: mesh
+      type(field_type)                   :: Ax, lumped_weight, res
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:322]: I don't think `res` and `MU` are used?

File: doc/user_guide/introduction_to_psykal.rst [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+  end module solver_mod
+
+
+.. _PSy-layer:
+
+PSy layer
+---------
+
+In the PSyKAl model, the PSy layer is the bridge between the Algorithm
+full field operations and the Kernel and Built-Ins individual element
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:345]: "full-field"

Comment (arporter) [reply doc/user_guide/introduction_to_psykal.rst:345]: and "Kernel/Built-Ins individual..."

File: doc/user_guide/introduction_to_psykal.rst [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+operations. As such, it is responsible for:
+
+1. calling any Kernel and expanding any Buit-In so that they iterate over
+   their specified interation space;
+2. map the Kernel and Built-In arguments supplied by an Algorithm ``invoke``
+   call to the arguments required by a Built-in or Kernel methods;
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:351]: "method;"

File: doc/user_guide/introduction_to_psykal.rst [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+3. include any required distributed-memory operations such as halo swaps
+   and reductions to guarantee the correctness of the code;
+4. and enable an entry point for the optimisation expert to provide
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:354]: s/and enable/providing/

File: doc/user_guide/introduction_to_psykal.rst [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+   additional optimisations for the target architecture.
+
+The PSy layer can be written manually but this is error prone and
+potentially complex to optimise. Therefore, the PSyclone code generation
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:358]: "code-generation"

File: doc/user_guide/introduction_to_psykal.rst [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+system automatically generates an initial version of the PSy layer by
+parsing the associated Algorithm file and each of the kernels used on it.
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:360]: s/on it/in it/

File: doc/user_guide/introduction_to_psykal.rst [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+This initial version of the PSy-layer can be further tuned with the PSyclone
+code transformation capabilities by providing a transformation script.
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:362]: "code-transformation"

File: doc/user_guide/introduction_to_psykal.rst [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+Each API comes with a set of specialise transformations designed for that API.
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:363]: "specialised"

File: doc/user_guide/introduction_to_psykal.rst [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+
+Also, in addition to all the functionality available for PSyIR Nodes, the
+PSy-layer nodes have a `dag()` method (standing for directed acyclic graph),
+outputs the PSyIR nodes and its data dependencies. By default a file in
```
Comment (arporter) [inline doc/user_guide/introduction_to_psykal.rst:367]: "which outputs...and their data dependencies."

File: doc/user_guide/index.rst [hunk 12]
```diff
@@ -82,38 +79,28 @@ PSyclone User Guide
     :maxdepth: 2
     :caption: Introduction
 
-    introduction
     getting_going
```
Comment (arporter) [inline doc/user_guide/index.rst:82]: In 'Getting Going', under 'run' we still assume that a PSyKAl DSL is being used:

image

Please could you update this text too?

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for taking this on Sergi, I think it's a big improvement. Both User Guide and Dev Guide build fine. Mostly just minor English corrections.

File: doc/user_guide/introduction_to_psykal.rst [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+Kernel layer and the Parallelisation System (PSy) layer; which together give
+the model its name.
+
+The Algorithm layer is responsible for providing a high-level description of the
+algorithm that the scientist wants to run. This layer operates on full fields
+and includes calls to kernels and built-ins.
+
+The Kernel layer contains the actual implementation of the kernels as functors.
+Each functor implements a method that operates on an individual section of the fields.
+Depending on the DSL, these can be a single elements, a vertical column, or a set of
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:58]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+vertical columns depending; the kernel also specifies metadata that describe the data
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:59]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+dependencies between different kernels. Bult-ins are similar to kernels but are
+provided by the PSyclone infrastructure itself.
+
+The PSy layer acts as a bridge between the algorithm and kernel layers. It is
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:63]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+responsible of enabling concurrent execution of kernels and to choose the optimal
+iteration order without breaking any of the kernel dependencies in order to
+maximise the performance for the target architecture. This layer can be tuned for
+specific platforms such as multi-node, multi-core and GPGPUs, without requiring
+any changes to the scientific algorithm or kernel layers. This allows for portable
+performance across different architectures.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:69]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+
+Rather than writing the PSy layer manually, the PSyclone code generation system uses
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:71]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+the provided Algorithm and Kernel implementations to generate an inital PSyIR implementation
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:72]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 19]
```diff
@@ -0,0 +1,408 @@
...
+of it, optionally with distributed-memory parallelism. This then can be programatically
+optimised by applying PSyclone transfomrations (e.g. kernel fusing, colouring,
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:74]: fixed

File: doc/user_guide/introduction_to_psykal.rst [hunk 20]
```diff
@@ -0,0 +1,408 @@
...
+inlining, ...) to better fit the target architecture.
+
+The rest of this page describes how to use ``psyclone`` command to process PSyKAl DSLs
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:77]: done

File: doc/user_guide/introduction_to_psykal.rst [hunk 21]
```diff
@@ -0,0 +1,408 @@
...
+Usage
+-----
+
+To use PSyclone to process a PSyKAl algorithm file, the ``-api API_NAME`` parameter
+must be provided. In addition, distributed memory can be switched on or off by 
+using the ``-dm``/``--dist_mem`` or ``-nodm``/``--no_dist_mem`` flags. For PSyKAl
+DSLs, the optional transformation script provided by the ``-s SCRIPT``
+parameter will be applied to the PSy-layer PSyIR representation.
+
+For example, the following command will process an LFRic PSyKAl algorithm file by
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:91]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 22]
```diff
@@ -0,0 +1,408 @@
...
+.. _psykal_usage:
+
+Usage
+-----
+
+To use PSyclone to process a PSyKAl algorithm file, the ``-api API_NAME`` parameter
+must be provided. In addition, distributed memory can be switched on or off by 
+using the ``-dm``/``--dist_mem`` or ``-nodm``/``--no_dist_mem`` flags. For PSyKAl
+DSLs, the optional transformation script provided by the ``-s SCRIPT``
+parameter will be applied to the PSy-layer PSyIR representation.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:89]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 23]
```diff
@@ -0,0 +1,408 @@
...
+
+   psyclone -api lfric -dm -s additional_optimisations.py algorithm.f90 \
+      -oalg algorithm_output.f90 -opsy psy_layer_output.f90
+
+To date, there are two PSyKAl DSL implementations: the
+:ref:`LFRic PSyKAl API <lfric-api>`, a mixed
+finite-element DSL used to implement the next-generation UK Met Office
+atmospheric model dynamical core; and the
+:ref:`GOcean PSyKAl API <gocean-api>`, a finite difference ocean model
+benchmark. Following sections provide more details about each specific API.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:104]: Ups. Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 24]
```diff
@@ -0,0 +1,408 @@
...
+Algorithm layer
+---------------
+
+As mentioned in the Introduction, the Algorithm layer provides a high-level
+description of the algorithm that the scientist would like to run, in terms
+of invokations to Kernel and Built-in operations. It operates on full fields
+and therefore it is not allowed to call the individual kernel executors nor
+include any parallelisation calls or directives. Instead, the algorithm layer
+uses the ``invoke`` subroutine, which takes as arguments one or more kernel
+functor constructors, as a Fortran type constructor, and optionally a name
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:119]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 25]
```diff
@@ -0,0 +1,408 @@
...
+argument. The kernel functors constructors, in turn, take as argument the full
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:120]: Removed "constructors" as I think its already clear.

File: doc/user_guide/introduction_to_psykal.rst [hunk 26]
```diff
@@ -0,0 +1,408 @@
...
+fields they operate on and any other value specified in its signature.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:121]: Yeah, done

File: doc/user_guide/introduction_to_psykal.rst [hunk 27]
```diff
@@ -0,0 +1,408 @@
...
+
+For example::
+
+    ...
+    call invoke(integrate_one_kernel(arg1,arg2), name="Example_Invoke")
+    ...
+
+A complete application can consist of many algorithm files, each of them
+contain as many ``invoke()`` calls as required. PSyclone is applied to
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:130]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 28]
```diff
@@ -0,0 +1,408 @@
...
+each individual algorithm layer file and must therefore be run multiple
+times if multiple algorithm files exist in a project.
+
+The algorithm developer is also able to reference more than one
+Kernel/Built-in within an invoke. In fact this feature is encouraged for
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:135]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 29]
```diff
@@ -0,0 +1,408 @@
...
+performance reasons. **As a general guideline the developer should aim to
+use as few invokes as possible with as many Kernel functors in them
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:137]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 30]
```diff
@@ -0,0 +1,408 @@
...
+as is possible**. The reason for this is that it allows for greater
+freedom for optimisations in the PSy layer as PSy layer optimisations
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:139]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 31]
```diff
@@ -0,0 +1,408 @@
...
+are limited to the contents of individual invokes - PSyclone
+currently does not attempt to optimise the PSy layer over multiple
+invoke calls.
+
+As well as generating the PSy layer code, PSyclone modifies the
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:144]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 32]
```diff
@@ -0,0 +1,408 @@
...
+Algorithm layer code, replacing ``invoke`` calls with calls to the
+generated PSy layer so that the algorithm code is compilable and
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:146]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 33]
```diff
@@ -0,0 +1,408 @@
...
+linkable to the PSy layer by adding in the appropriate ``use``
+statement. For example, the invoke above is translated into something
+like the following::
+
+  ...
+  use psy, only : invoke_example_invoke
+  ...
+  call invoke_example_invoke(arg1,arg2)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:154]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 34]
```diff
@@ -0,0 +1,408 @@
...
+  ...
+
+The ``name`` argument in the invoke call is optional. If supplied it
+must be a string literal, labels are not case-sensitive and must be
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:158]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 35]
```diff
@@ -0,0 +1,408 @@
...
+valid Fortran names (e.g. ``name="compute(1)"`` is invalid).
+The label is used to name the corresponding PSy-layer routine generated
+by PSyclone in other to make debugging and profiling outputs more readable.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:161]: Done

File: doc/developer_guide/psykal.rst [hunk 36]
```diff
@@ -31,18 +32,23 @@
...
     SOURCE_FILE = ("../../src/psyclone/tests/test_files/"
         "gocean1p0/test11_different_iterates_over_one_invoke.f90")
 
 
-Parsing Code (new approach)
-###########################
+Parsing DSL Code (new approach)
+===============================
 
 The new approach to modifying existing code is to first parse it into
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:53]: Removed

File: doc/developer_guide/psykal.rst [hunk 37]
```diff
@@ -103,8 +109,8 @@ yet used by the the rest of PSyclone (see `generator.py` for the
 relevant GOcean code and prototype LFRic code).
 
 
-Parsing Code (original approach)
-################################
+Parsing DSL code (original approach)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:112]: Done

File: doc/developer_guide/psykal.rst [hunk 38]
```diff
@@ -103,8 +109,8 @@ yet used by the the rest of PSyclone (see `generator.py` for the
...
+====================================
 
 The original way to parse code is to use the PSyclone `parse` module
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:115]: Done

File: doc/developer_guide/psykal.rst [hunk 39]
```diff
@@ -364,3 +370,76 @@ instance has been created (by passing it an `fparser1` parse tree) it can
 return information about the metadata contained therein. Moving from
 `fparser1` to `fparser2` would required changing the parse code logic
 in each of the API-specific classes.
+
+The PSy-Invokes-Invoke-InvokeSchedule tree
+==========================================
+
+The PSy-layer of a single algorithm file is represented by the **PSy** class.
+The PSy class has an **Invokes** object which contain one or more
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:378]: Done

File: doc/developer_guide/psykal.rst [hunk 40]
```diff
@@ -364,3 +370,76 @@ instance has been created (by passing it an `fparser1` parse tree) it can
...
+**Invoke** instances (one for each invoke in the algorithm layer).
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:379]: Done

File: doc/developer_guide/psykal.rst [hunk 41]
```diff
@@ -364,3 +370,76 @@ instance has been created (by passing it an `fparser1` parse tree) it can
...
+Each **Invoke** has an **InvokeSchedule** object with the PSyIR tree
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:380]: Done

File: doc/developer_guide/psykal.rst [hunk 42]
```diff
@@ -364,3 +370,76 @@ instance has been created (by passing it an `fparser1` parse tree) it can
...
+that describes the PSy layer invoke subroutine.
+This subroutine is called by the Algorithm layer and itself calls one or
+more kernels and/or implements any required Built-in operations.
+
+All this classes can be specialised in each PSyclone API to support the
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:385]: Done

File: doc/developer_guide/psykal.rst [hunk 43]
```diff
@@ -364,3 +370,76 @@ instance has been created (by passing it an `fparser1` parse tree) it can
...
+specific features of the APIs. The class diagram for the above base classes
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:386]: Done

File: doc/developer_guide/psykal.rst [hunk 44]
```diff
@@ -364,3 +370,76 @@ instance has been created (by passing it an `fparser1` parse tree) it can
...
+is shown below using the dynamo0.3 API as an illustration. This class diagram
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:387]: Done

File: doc/developer_guide/psykal.rst [hunk 45]
```diff
@@ -364,3 +370,76 @@ instance has been created (by passing it an `fparser1` parse tree) it can
...
+========================================
+
+To add a new built-in operation into an API use the following steps:
+
+ 1. Identify the PSyclone source file for the API to be extended. *e.g.* for
+    the LFRic API it is ``src/psyclone/domain/lfric/lfric_builtins.py``.
+ 2. Edit this source file to create the class for this new call. It must
+    inherit from the API-specific parent class for Built-in operations
+    (``LFRicBuiltInKern`` for the LFRic API).
+ 3. Implement ``__str__`` and ``gen_code()`` methods for this new class.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/developer_guide/psykal.rst:412]: Updated

File: doc/user_guide/index.rst [hunk 46]
```diff
@@ -44,35 +44,32 @@
...
-    `Reference Guide <https://psyclone-ref.readthedocs.io/>`_.
+PSyclone is a source-to-source Fortran compiler designed to programmatically
+optimise, parallelise and instrument HPC applications via user-provided
+transformation scripts.
+
+By encapsulating the performance-portability aspects (e.g. whether to
+parallelise with OpenMP or OpenACC), these scripts enable a separation of
+concerns between the scientific implementation and the optimisation choices.
+This allows each aspect to be explored and developed largely independently.
+Additionally, PSyclone supports the development of kernel-based Fortran-embedded
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/index.rst:55]: Done

File: doc/user_guide/psyke.rst [hunk 47]
```diff
@@ -153,12 +153,11 @@ are used or not.
...
 
-As noted in the :ref:`distributed_memory` section, support for distributed
-memory in PSyclone is currently limited to the
-:ref:`LFRic API <lfric-api>`. Since the implementation
-generates calls to LFRic infrastructure (e.g. runtime checks for status
-of field halos), code extraction is not allowed when distributed memory
-is enabled.
+As noted in the :ref:`PSyKAl Introduction <introduction_to_psykal>` section,
+support for distributed memory in PSyclone is currently limited to the PSyKAl
+APIs. Since the PSyKE implementation generates calls to infrastructure (e.g.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/psyke.rst:158]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 48]
```diff
@@ -0,0 +1,408 @@
...
+So, for the above example, the generated PSy-layer subroutine
+will be named "invoke_example_invoke", otherwise a numeric index is given.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:163]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 49]
```diff
@@ -0,0 +1,408 @@
...
+
+
+.. _kernel-layer:
+
+Kernel layer
+------------
+
+In the PSyKAl model, the Kernel code operates on an individual element of
+a field (such as a column of cells). The reason for doing this is that it
+gives the PSy layer flexibility of chosing the iteration order and spacial
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:187]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 50]
```diff
@@ -0,0 +1,408 @@
...
+domain parallelisation can be exploited. The Kernel layer is not allowed
+to include any parallelisation calls or directives and works on
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:189]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 51]
```diff
@@ -0,0 +1,408 @@
...
+raw Fortran arrays (to allow the compiler to optimise the code).
+Since a Kernel is called over the spatial domain (by the PSy layer) it
+must take at least one field or operator as an argument.
+
+Kernels are implemented as Fortran Fuctors. Functors are objects that can be
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:194]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 52]
```diff
@@ -0,0 +1,408 @@
...
+treated as if they are functions. As such they have two main interfaces for
+calling and providing arguments to them: the object constructor (used by the
+algorithm layer) and a method that executes the code of the functor (used by
+the PSy-layer).
+
+PSyKal applications accept one or more modules providing kernels, each of which
+can contain one or more kernel functors. Each kernel functor provides a set of
+meta-data attributes and a method with the same name as the functor with the
+prefix ``_code```.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:203]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 53]
```diff
@@ -0,0 +1,408 @@
...
+
+In the example below the module ``w3_solver_kernel_mod`` contains one kernel
+named ``w3_solver_kernel_type`` and its individual element execution method in
+subroutine ``w3_solver_code``.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:207]: Ok deleted previous sentence.

File: doc/user_guide/introduction_to_psykal.rst [hunk 54]
```diff
@@ -0,0 +1,408 @@
...
+
+The metadata is API-specific and describes the kernel iteration space and
+dependencies, so that PSyclone can generate correct PSy layer code. The
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:210]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 55]
```diff
@@ -0,0 +1,408 @@
...
+metadata is provided by the kernel developer, whom it has to guarantee its
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:211]: Done

File: doc/user_guide/introduction_to_psykal.rst [hunk 56]
```diff
@@ -0,0 +1,408 @@
...
+      procedure, nopass :: solver_w3_code
+    end type
+  
+  contains
+  
+    subroutine solver_w3_code(nlayers,                                 &
+                              x, rhs,                                  &
+                              chi_1, chi_2, chi_3, ascalar,            &
+                              ndf_w3, undf_w3, map_w3, w3_basis,       &
+                              ndf_w0, undf_w0, map_w0, w0_diff_basis,  &
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:242]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 1 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:277]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 3 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:322]: Removed

File: doc/user_guide/introduction_to_psykal.rst (hunk 2 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:300]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 4 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:345]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 5 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:351]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 6 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:354]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 7 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:358]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 8 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:360]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 10 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:363]: Done

File: doc/user_guide/introduction_to_psykal.rst (hunk 11 above)
Comment (sergisiso) [reply doc/user_guide/introduction_to_psykal.rst:367]: Done

File: doc/user_guide/index.rst (hunk 12 above)
Comment (sergisiso) [reply doc/user_guide/index.rst:82]: I just made a light update on this to make clear that we are talking about PSyKAl here. But I am planning to revisit this whole section in a separate PR.

**Comment by sergisiso** [comment]
@arporter This is ready for another review

File: doc/user_guide/introduction_to_psykal.rst (hunk 49 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/introduction_to_psykal.rst]: I suggest we drop "spatial" (or keep it and correct the spelling).

File: doc/user_guide/introduction_to_psykal.rst (hunk 4 above)
Comment (arporter) [reply doc/user_guide/introduction_to_psykal.rst]: Almost, should be "full-field", not "full-fields".

File: doc/user_guide/index.rst (hunk 12 above)
Comment (arporter) [reply doc/user_guide/index.rst:82]: Great, thanks.

File: doc/user_guide/introduction_to_psykal.rst (hunk 49 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/introduction_to_psykal.rst]: I've fixed this too.

**Review by arporter — APPROVED** [review]
All requested changes have been made, thanks.
I've taken the liberty of fixing a couple of remaining typos.
Both Dev and User Guides build fine.
Will proceed to merge.

**Pull request merged**